### PR TITLE
[SPARK-30201][SQL] HiveOutputWriter standardOI should use ObjectInspectorCopyOption.DEFAULT

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Now spark use `ObjectInspectorCopyOption.JAVA` as oi option which will convert any string to UTF-8 string. When write non UTF-8 code data, then `EFBFBD` will appear.
We should use `ObjectInspectorCopyOption.DEFAULT` to support pass the bytes.

### Why are the changes needed?

Here is the way to reproduce:
1. make a file contains 16 radix 'AABBCC' which is not the UTF-8 code.
2. create table test1 (c string) location '$file_path';
3. select hex(c) from test1; // AABBCC
4. craete table test2 (c string) as select c from test1;
5. select hex(c) from test2; // EFBFBDEFBFBDEFBFBD

### Does this PR introduce any user-facing change?

No.

### CODE REVIEW

**Comment by ulysses-you** [comment]
ping @cloud-fan @srowen . I think it's your area.

**Comment by srowen** [comment]
This isn't really my area, but I don't quite understand how this arises in practice? what value does this hex string encode, and how would Spark write it? Spark and Hadoop generally always encode strings as UTF-8

**Comment by ulysses-you** [comment]
> This isn't really my area, but I don't quite understand how this arises in practice? what value does this hex string encode, and how would Spark write it? Spark and Hadoop generally always encode strings as UTF-8
The hex string is just to make reproduce easily.

This is my scene:
1. We write file using hadoop api to write bytes which is not utf-8 code and craete hive table with that location. 
2. We use spark sql read the table and write into an another table.

**Comment by srowen** [comment]
I don't know how to evaluate the other effects of this change -- what are they? What does ObjectInspectorCopyOption do?

Also I don't know a lot about this, but it doesn't look like you're writing a table; you're writing some bytes? how is this read?

**Comment by ulysses-you** [comment]
Simply, We can find more details in `HiveInspectors`. The important difference between `DEFAULT` and `JAVA` is `preferWritable()` that decide to make `Text` or `String`. As default hive serde, spark use `LazySimpleSerDe` which preferWritable is always true.

**Comment by cloud-fan** [comment]
So spark can read `test1` correctly, but when we write `test1` to `test2`, the data is corrupted?

**Comment by ulysses-you** [comment]
Yes.  It is easy to reproduce.

**Comment by cloud-fan** [comment]
can you point to me the difference between Spark reading a hive table and Spark reading a hive table then writing to another table? Is the bug in data reading or data writing?

**Comment by ulysses-you** [comment]
The problem is writing.  When a column type is string,  spark will read bytes to UTF8String.  This step not actually check the UTF-8 code,   just copy bytes.  
Then convert the UTF8String.toString during write. This step will convert every bytes as UTF-8  string. 
As the result,  non UTF-8 code bytes will error. 

So we should pass bytes directly without tostring in right sence.

**Comment by srowen** [comment]
Hm, I'm not sure about this. Converting bytes to Text would always use UTF-8, yes. Converting bytes to String would need an encoding, which I assume somewhere defaults to a platform encoding, right? or else how is that defined?

But back to my question, how do you get non-UTF-8 data here in the first place? This might be a dumb question, I don't know how other writers might work. Is the case a table where the serde properties define the encoding to something besides UTF-8? Because here you are just writing bytes directly, which is why I ask how this arises in practice. But in that case does this not get handled properly somewhere else by Hive or Spark?

I guess I'm saying I don't know if this demonstrates a real-world issue, vs what happens if you hack in some bytes.

**Comment by cloud-fan** [comment]
> how do you get non-UTF-8 data here in the first place?

This is a good question. once data is load into Spark, string type must be a UTF-8 encoded binary. If it's not, then data loading (table reading) has a bug.

It's a little weird to hear that: hive table reading is fine, but the writing after reading is problematic because string is not UTF8 encoded. Seems like the table reading is wrong at the first place (produce non-utf8-encoded string).

**Comment by ulysses-you** [comment]
> how do you get non-UTF-8 data here in the first place?

As I said we use hadoop api to write bytes  and create table with it. In normal this bytes are UTF8 code, but sometimes it does not. It is hard to say why it happened because data are historical.

Then we use sha1 or md5 to check if data is correct which function also use bytes directly. And we meet this problem.

**Comment by ulysses-you** [comment]
Actually, we also do the same option in hive, and hive is correct. Then we find the problem is spark use JAVA option which is fixed option. 

The root cause is the JAVA option always convert `UTF8String` to string instead of convert `UTF8String` to `Text`.

BTW Hive `Text` and Spark `UTF8String` has the same processing mode with UTF8 code that copy bytes directly and to UTF8 String lazily(invoke toString or other method).

**Comment by srowen** [comment]
I still don't really understand the problem here. You wrote some legacy data but with no indication of its encoding somehow? How does Hive understand it correctly, if it's not specified in serde properties, etc?

**Comment by ulysses-you** [comment]
E.G. A Text table. Hive just read it into `Text`. When we select and print, it will to string and show somethings we can not readable.
But Write data use the origin bytes and does not care what code it is. So we should avoid convert the bytes to string during writing.

**Comment by ulysses-you** [comment]
Now spark write pipeline is:
Non UTF8 bytes -> UTF8String.fromBytes() -> toString() -> to Text -> write.

But the correct write pipeline is:
Non UTF8 bytes -> UTF8String.fromBytes() -> getBytes() -> to Text -> write

**Comment by srowen** [comment]
The second pipeline is definitely wrong. Construing bytes that encode a string in some non-UTF8 encoding as UTF8, then getting the bytes back, is going to be incorrect in some cases.

I am still not clear how you supply bytes that encode a string in something other than UTF8 to Spark, but tell it to interpret as UTF8?

**Comment by ulysses-you** [comment]
No. UTF8String.fromBytes() dose not care what code it is, this option just copy bytes. And getBytes() also use the origin bytes without any code.
So the second pipeline just bytes to bytes.

**Comment by cloud-fan** [comment]
I think it's about converting UTF8String to hive Text. `getBytes() -> to Text -> write` seems corrected to me.

**Comment by ulysses-you** [comment]
Yes. It is that case.
Change to `DEFAULT` is also more suitable with `HiveInspectors` than `JAVA`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala [hunk 1]
```diff
@@ -133,7 +133,7 @@ class HiveOutputWriter(
   private val standardOI = ObjectInspectorUtils
     .getStandardObjectInspector(
       tableDesc.getDeserializer(jobConf).getObjectInspector,
-      ObjectInspectorCopyOption.JAVA)
+      ObjectInspectorCopyOption.DEFAULT)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala]: Can you point to me where we convert `UTF8String` to string or bytes according to this mode? `UTF8String` is a Spark class and I assume this part of code should be in Spark?

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala]: The code is at `HiveInspectors.wrapperFor()`.
```
case _: StringObjectInspector if x.preferWritable() =>
        withNullSafe(o => getStringWritable(o))
case _: StringObjectInspector =>
        withNullSafe(o => o.asInstanceOf[UTF8String].toString())
```

Before write data with hiveWriter, spark will use `HiveInspectors.wrapperFor()` to convert `UTF8String` to `Text` or `String`.

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala]: ah I see. Can you add some comments here? Saying that the `DEFAULT` mode indicates Spark to create Text with the bytes of `UTF8String`.

**Comment by cloud-fan** [comment]
I think this is a malformed use case. The hive table returns a malformed `UTF8String` (the bytes is not a utf8 encoded string), and we want to write that malformed `UTF8String` to another hive table.

I'm OK with the fix here, as an improvement of hive compatibility. We should write the malformed `UTF8String` to hive table as it is.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveFileFormat.scala]: OK. I will add it later.

**Comment by srowen** [comment]
I can sort of buy this logic if it only affects Hive and basically doesn't make it work any _less_. But the round trip from non-UTF-8-bytes -> read as UTF-8 -> get bytes will definitely not give you back the same bytes in some cases. It won't parse in some cases either. This seems a bit fragile.

The change is also broader than just affecting this case. It affects much more, potentially, about how everything is read. It'll be interesting to see how tests run, but this makes me a bit nervous.

**Comment by srowen** [comment]
Those build failures are spurious. @shaneknapp this is another example of what I was talking about.

**Comment by shaneknapp** [comment]
test this please

**Comment by ulysses-you** [comment]
It seems like `HiveInspectors` miss some check, `HiveVarchar` and `HiveChar`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala [hunk 2]
```diff
@@ -305,12 +305,17 @@ private[hive] trait HiveInspectors {
         withNullSafe(o => getByteWritable(o))
       case _: ByteObjectInspector =>
         withNullSafe(o => o.asInstanceOf[java.lang.Byte])
-      case _: JavaHiveVarcharObjectInspector =>
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: do we still need to handle `JavaHiveVarcharObjectInspector`?

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: Sorry for late reply. It is no need. `case _: HiveVarcharObjectInspector =>` has already include this.

I don not known why some type add additional java check like `JavaHiveDecimalObjectInspector`, `JavaDateObjectInspector`. But most type has not, like `String`, `Int`, `Binary`.

So shall we retain `JavaHiveVarcharObjectInspector ` check to respect ?

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: do you mean `JavaHiveVarcharObjectInspector` extends `HiveVarcharObjectInspector`?

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: Yes.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
Thanks!

**Comment by anuragmantri** [comment]
@ulysses-you @cloud-fan - I verified this issue is present in branch-2.4. (By running the unit test and observing the failure.) 
`[info]   == Results ==
[info]   !== Correct Answer - 1 ==   == Spark Answer - 1 ==
[info]   !struct<>                   struct<hex(c):string>
[info]   ![AABBCC]                   [EFBFBDEFBFBDEFBFBD] (QueryTest.scala:163)
[info]   org.scalatest.exceptions.TestFailedException:`

I created a PR for backport into 2.4 branch - https://github.com/apache/spark/pull/29948 could you please take a look? There were no conflicts in the cherry-pick (except for import ordering). Thanks.

CC: @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Thank you, @anuragmantri !

**Pull request closed without merging**